sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2018,6 +2018,8 @@ class Analyzer(
             fail(child, DateType, walkedTypePath)
           case (StringType, to: NumericType) =>
             fail(child, to, walkedTypePath)
+          case (from: ArrayType, to: ArrayType) if !from.containsNull =>
```

Comment:
I will try improving the `SimplifyCasts` rule to force to eliminate the cast from non-element-nullable array to nullable ones. I do not understand the following. Will it be automatically done by improving the `SimplifyCasts` or do we need to improve another rule?
> "we can handle map too"

I will add unit tests for it. I think that it is good to add a benchmark to show degree of improvements.